File: aggregated_dependency_score.py
import logging
from math import log


MIN_TRUSTWORTHINESS = 0.8
# this is noted as "k" in the paper
TRUSTWORTHINESS_OFFSET_PARAM = 60
# this is noted as "e" in the paper
TRANSITIVE_TRUSTWORTHINESS_POWER = 0.5


def trustworthiness_from_score(score: float) -> float:
    """this function is noted as "f" in the paper"""
    return 1 - (1 - MIN_TRUSTWORTHINESS) * (
        1 - log(1 + (TRUSTWORTHINESS_OFFSET_PARAM - 1)*score)
        / log(TRUSTWORTHINESS_OFFSET_PARAM)
    )


def score_from_trustworthiness(trustworthiness: float) -> float:
    """this function is noted as $\\tilde{f}^{-1}$ in the paper"""

    t = trustworthiness
    k = TRUSTWORTHINESS_OFFSET_PARAM

    if t < MIN_TRUSTWORTHINESS:
        return 0

    return (
        ( 1 - k**(1 - (1-t)/0.2) )
        / (1 - k)
    )

def compute_aggregrated_trustworthiness(
    package,
    get_direct_dependencies,
    compute_intrinsic_score,
):

    s = compute_intrinsic_score(package)
    t = trustworthiness_from_score(s)

    logging.debug(f"intrinsic trustworthiness for {package}: {t} (score: {s})")

    t_hat = compute_transitive_trustworthiness(
        package,
        get_direct_dependencies,
        compute_intrinsic_score,
    )

    logging.debug(f"transitive trustworthiness for {package}: {t_hat}")

    return t * t_hat


def compute_transitive_trustworthiness(
    package,
    get_direct_dependencies,
    compute_intrinsic_score,
):
    e = TRANSITIVE_TRUSTWORTHINESS_POWER

    result = 1

    for dep in get_direct_dependencies(package):
        t_prime = compute_aggregrated_trustworthiness(
            dep,
            get_direct_dependencies,
            compute_intrinsic_score,
        )

        logging.debug(f"aggregated trustworthiness for {dep} (dependency of {package}): {t_prime}")

        t_hat = compute_transitive_trustworthiness(
            dep,
            get_direct_dependencies,
            compute_intrinsic_score,
        )

        logging.debug(f"transitive trustworthiness for {dep} (dependency of {package}): {t_hat}")

        result *= t_prime * t_hat**e

    return result

File: aggregated-dependency-score.go
package aggregdepscore

import (
	"context"
	"fmt"
	"math"
)

// transitiveTrustworthinessExponent is noted as "e" in the design paper
const transitiveTrustworthinessExponent = 1.5

type Package struct {
	Ecosystem string
	Name      string
	Version   string
}

func (e *Evaluator) EvaluateScore(ctx context.Context, p Package) (float64, error) {
	aggregatedTrustworthiness, err := e.trustworthiness.evaluate(ctx, p, nil)
	if err != nil {
		return 0.0, err
	}

	return e.converter.ScoreFromTrustworthiness(aggregatedTrustworthiness), nil
}

func NewEvaluator(intrinsic IntrinsicTrustworthinessEvaluator, deps DependencyResolver) (*Evaluator, error) {
	if intrinsic == nil {
		return nil, fmt.Errorf("intrinsic trustworthiness evaluator is required")
	}

	if deps == nil {
		return nil, fmt.Errorf("dependency resolver is required")
	}

	return &Evaluator{
		trustworthiness: trustwhorthinessEvaluator{
			intrinsic: intrinsic,
			deps:      deps,
		},
		converter: &DefaultScoreTrustworthinessConverter{},
	}, nil
}

type Evaluator struct {
	trustworthiness trustwhorthinessEvaluator
	converter       ScoreTrustworthinessConverter
}

type trustwhorthinessEvaluator struct {
	intrinsic IntrinsicTrustworthinessEvaluator
	deps      DependencyResolver
}

type IntrinsicTrustworthinessEvaluator interface {
	EvaluateIntrinsicTrustworthiness(ctx context.Context, p Package) (float64, error)
}

type DependencyResolver interface {
	GetDirectDependencies(ctx context.Context, p Package) ([]Package, error)
}

func (evaluator *trustwhorthinessEvaluator) evaluate(ctx context.Context, p Package, ancestors map[string]struct{}) (float64, error) {
	intrinsic, err := evaluator.intrinsic.EvaluateIntrinsicTrustworthiness(ctx, p)
	if err != nil {
		return 0.0, fmt.Errorf("evaluating intrinsic trustworthiness of package: %w", err)
	}

	result := intrinsic

	deps, err := evaluator.deps.GetDirectDependencies(ctx, p)
	if err != nil {
		return 0.0, fmt.Errorf("getting direct dependencies of package: %w", err)
	}

	for _, dep := range deps {
		// XXX sometimes different names can refer to the same package,
		// for instance with gopkg.in URLs;
		// XXX should we consider the version as well?
		if _, ok := ancestors[dep.Name]; ok {
			// depedency cycle
			// TODO emit a log
			continue
		}

		// copy ancestors to avoid modifying the original map;
		// one day we may want to run the algorithm in parallel
		// so we will need to be careful with shared state
		childAncestors := make(map[string]struct{})
		for k, v := range ancestors {
			childAncestors[k] = v
		}
		childAncestors[p.Name] = struct{}{}

		tPrimeQ, err := evaluator.evaluate(ctx, dep, childAncestors)
		if err != nil {
			return 0.0, fmt.Errorf("evaluating aggregated trustworthiness of %s: %w", dep.Name, err)
		}

		result *= math.Pow(tPrimeQ, transitiveTrustworthinessExponent)
	}

	return result, nil
}
